lib/spack/spack/binary_distribution.py

```diff
@@ -325,24 +332,36 @@ def update(self):
 
         fetch_errors = []
         all_methods_failed = True
+        ttl = spack.config.get("config:binary_index_ttl", 600)
+        now = time.time()
 
         for cached_mirror_url in self._local_index_cache:
             cache_entry = self._local_index_cache[cached_mirror_url]
             cached_index_hash = cache_entry["index_hash"]
             cached_index_path = cache_entry["index_path"]
             if cached_mirror_url in configured_mirror_urls:
-                # May need to fetch the index and update the local caches
-                try:
-                    needs_regen = self._fetch_and_cache_index(
-                        cached_mirror_url, expect_hash=cached_index_hash
-                    )
+                # Only do a fetch if the last fetch was longer than TTL ago
+                if (
+                    ttl > 0
+                    and cached_mirror_url in self._last_fetch_times
+                    and now - self._last_fetch_times[cached_mirror_url] < ttl
```

Comment:
I'd like to add another option to the method signature for `update`:

```
def update(self, with_cooldown=False)
```

and update this like

```
               if (
                     with_cooldown
                     and ttl > 0
                     and cached_mirror_url in self._last_fetch_times
                     and now - self._last_fetch_times[cached_mirror_url] < ttl
```

in which case `find_by_hash` can set `with_cooldown=True`.

My goal is to avoid cases where users update the cache (in their program), then immediately call `update` (I think they would be confused if `update` was a noop in come circumstances).

I considered requesting moving the timer out of here, but since the timers are on a per-mirror-url basis, that would be awkward to do.

Reply:
> My goal is to avoid cases where users update the cache (in their program), then immediately call `update` (I think they would be confused if `update` was a noop in come circumstances).

By "in their program," you mean in a Spack extension? This cooldown period only acts within each Spack invocation, so if they're using the command line there won't be a noticeable difference.

Reply:
> By "in their program," you mean in a Spack extension?

Yes, or in some script they are running with `spack-python`

Reply:
Done.